### PR TITLE
cmake: add RGW and MDS to libcephd

### PR DESCRIPTION
**Pull request by bassam**
This PR adds RGW and MDS to embedded Ceph. The approach is similar to the work to add MON and OSD keeping the changes to a minimum. Also in here is a commit to make the FCGI dependency optional

### CODE REVIEW

File: src/rgw/rgw_main.cc [hunk 1]
```diff
@@ -110,7 +110,9 @@ static void signal_fd_finalize()
 static void handle_sigterm(int signum)
 {
   dout(1) << __func__ << dendl;
+#if defined(WITH_RADOSGW_ASIO_FRONTEND)
```
Comment (cbodley) [inline src/rgw/rgw_main.cc]: ASIO -> FCGI

File: src/rgw/rgw_main.cc [hunk 2]
```diff
@@ -316,7 +322,9 @@ int main(int argc, const char **argv)
   
   curl_global_init(CURL_GLOBAL_ALL);
   
+#if defined(WITH_RADOSGW_ASIO_FRONTEND)
```
Comment (cbodley) [inline src/rgw/rgw_main.cc]: ASIO -> FCGI

**Comment by bassam** [comment]
@cbodley fixed in rebased bb0298f

**Comment by bassam** [comment]
@cbodley @tchaikov can you please take a look. I just rebased and fixed conflicts. Would be great if we can put this in master before more conflicts arise :-)

**Comment by cbodley** [comment]
the code looks good. does it actually run and pass tests?

**Comment by tchaikov** [comment]
@bassam @cbodley i will pull this into my test branch to see if it builds on shaman and passes rados qa tests.

**Comment by cbodley** [comment]
what i meant was, do we have any test plans for the embedded versions? for example, can it successfully bring up two functional daemons if you `cephd_run_*()` them from separate threads? i'm guessing that we'll run into a lot of problems with g_ceph_context and other global state, and someone will need to do that testing and work through the issues. until that happens, i think it's important to document these constraints in `libcephd.h` (and possibly enforce them by having the `cephd_run_*()` functions immediately return an error if another daemon is already running)

**Comment by tchaikov** [comment]
> do we have any test plans for the embedded versions?

no, AFAICT. at least, not yet.

**Comment by bassam** [comment]
@cbodley there are not tests yet. `cephd_run_*()` does run well when there is a single daemon per process as you observed, but not for multiple daemons per process. The plan is to keep driving this work such that it can be used in true embedded form.

Note also that embedded ceph is not currently included in distro packages. Once the work is complete we would need to do so, add documentation, and integration testing.

I'll add some documentation on the current constraints in a separate PR.

**Pull request merged**